### PR TITLE
Add full support for JSON-LD and proofs

### PR DESCRIPTION
**Pull request by stevenvegt**
This PR contains:

* Support for detached payload signatures in crypto
* The correct  way of signing JSON-LD documents
  * Stored and cached JSON-LD contexts
  * Utils to canonicalize JSON-LD documents  
  * Abstraction and backwards compatibility with the legacy proof
* Partial implementation of the verifier role

TODO:
* [x] Do not resolve unknown JSON-LD contexts when in `strict` mode
* [x] Sign new transactions with the legacy proof

Note: The vcr/api/v2 verifier is not yet complete. It now only checks the signature and times. In the #825 PR credential revocation is added.

### CODE REVIEW

File: vcr/signature/ldutils.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	return nil, ld.NewJsonLdError(ld.LoadingDocumentFailed, nil)
+}
+
+// NewContextLoader creates a new JSON-LD context loader with the embedded FS as first loader.
+// It loads the most used context from the embedded FS. This ensures the contents cannot be altered.
+// If allowExternalCalls is set to true, it also loads external context from the internet.
+func NewContextLoader(allowExternalCalls bool) (ld.DocumentLoader, error) {
+	var nextLoader ld.DocumentLoader
+	if allowExternalCalls {
+		nextLoader = ld.NewDefaultDocumentLoader(nil)
```
Comment (woutslakhorst) [inline vcr/signature/ldutils.go:61]: untested

**Review by woutslakhorst — CHANGES_REQUESTED** [review]
- missing copyrights

File: vcr/assets/assets/contexts/lds-jws2020-v1.ldjson [hunk 2]
```diff
@@ -0,0 +1,53 @@
+{
```
[the remark this answers is not in the document]
Comment (stevenvegt) [reply vcr/assets/assets/contexts/lds-jws2020-v1.ldjson:1]: Afaik you cannot add comments to json and the json-ld spec does not have a comment field or equivalent.

File: vcr/assets/assets/contexts/nuts.ldjson [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
```
[the remark this answers is not in the document]
Comment (stevenvegt) [reply vcr/assets/assets/contexts/nuts.ldjson:1]: I can make it empty and add actual data in further PRs.

File: vcr/signature/proof/proof.go [hunk 4]
```diff
@@ -19,12 +19,68 @@
...
+
+// UnmarshalProofValue unmarshalls the signature of the document in the provided target
+func (d SignedDocument) UnmarshalProofValue(target interface{}) error {
+	asJSON, err := json.Marshal(d["proof"])
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(asJSON, target)
+}
+
```
[the remark this answers is not in the document]
Comment (stevenvegt) [reply vcr/signature/proof/proof.go:69]: Not sure: the SignedDocument is currently not really a thing in the VC standards
Perhaps if we want to extend the go-did library with cryptographic operation support.

File: vcr/verifier/verifier.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+		}
+		return legacyProof.Verify(signedDocument.DocumentWithoutProof(), signature.LegacyNutsSuite{}, pk)
+	}
+	return err
+
+}
+
+// Verify implements the verify interface.
+// It currently checks if the credential has the required fields and values, if it is valid at the given time and optional the signature.
+// TODO: check for revoked credentials.
```
[the remark this answers is not in the document]
Comment (stevenvegt) [reply vcr/verifier/verifier.go:128]: The v1 api currently has this logic implemented in the vcr.go. These are needed for the v2 api to be complete and are part of other PRs.

**Review by woutslakhorst — APPROVED** [review]

**Pull request merged**